src/Data/LLVM/BitCode/GetBits.hs

```diff
@@ -125,112 +117,233 @@ instance MonadPlus GetBits where
+-- unlifted values and avoids the final shift by being smarter about individual
+-- compositions.  Their functionality should be identical, but it may be easier
+-- to debug the first.
+
+-- extractFromByteString bitLimit startBit numBits bs =
+--   let Bytes' s8 = fst (bitsToBytes startBit)
+--       Bytes' r8 = fst (bitsToBytes numBits)
+--       rcnt = r8 + 2 -- 2 == pre-shift overflow byte on either side
+
+--       -- Extract the relevant bits from the ByteCode, with padding to byte
+--       -- boundaries into ws.
+--       ws = BS.take rcnt $ BS.drop s8 bs
+
+--       -- Combine the extracted bytes into an Integer value in wi.
+--       wi = BS.foldr (\w a -> a `shiftL` 8 .|. fromIntegral w) (0::Int) ws
+
+--       -- Mask is 0-bit based set of bits wanted in the result
+--       mask = ((1::Int) `shiftL` bitCount numBits) - 1
+
+--       -- Shift the desired value down to byte alignment and then discard any
+--       -- excess high bits.
+--       vi = wi `shiftR` (bitCount startBit .&. 7) .&. mask
+
+--       updPos = addBitCounts startBit numBits
+--   in if updPos > bitLimit
+--      then Left ("Attempt to read bits past limit (newPos="
+--                 <> show updPos <> ", limit=" <> show bitLimit <> ")"
+--                )
+--      else Right (vi, updPos)
+
+extractFromByteString :: Int# {-^ the last bit accessible in the ByteString -}
+                      -> Int# {-^ the bit to start extraction at -}
+                      -> Int# {-^ the number of bits to extract -}
                       -> ByteString {-^ the ByteString to extract from -}
-                      -> Either String (Integer, NumBits)
-extractFromByteString bitLimit startBit numBits bs =
-  let Bytes' s8 = fst (bitsToBytes startBit)
-      Bytes' r8 = fst (bitsToBytes numBits)
-      rcnt = r8 + 2 -- 2 == pre-shift overflow byte on either side
-
-      -- Extract the relevant bits from the ByteCode, with padding to byte
-      -- boundaries into ws.
-      ws = BS.take rcnt $ BS.drop s8 bs
-
-      -- Combine the extracted bytes into an Integer value in wi.
-      wi = BS.foldr (\w a -> a `shiftL` 8 .|. fromIntegral w) (0::Integer) ws
-
-      -- Mask is 0-bit based set of bits wanted in the result
-      mask = ((1::Integer) `shiftL` bitCount numBits) - 1
-
-      -- Shift the desired value down to byte alignment and then discard any
-      -- excess high bits.
-      vi = wi `shiftR` (bitCount startBit .&. 7) .&. mask
+                      -> Either String (() -> (# Int#, Int# #))
+extractFromByteString !bitLim# !sBit# !nbits# bs =
+     if isTrue# ((1# `uncheckedIShiftL#` (nbits#)) /=# 0#)
+        -- (nbits# -# 1#) above would allow 64-bit value extraction, but this
+        -- function cannot actually support a size of 64, because Int# is signed,
+        -- so it doesn't properly use the high bit in numeric operations.  This
+        -- seems to be OK at this point because LLVM bitcode does not attempt to
+        -- encode actual 64-bit values.
```

Comment:
I'm not sure what you mean by "LLVM bitcode does not attempt to encode actual 64-bit values". After all, `i64` is most definitely a type in LLVM, which `(u)int64_t` compiles down to.

Reply:
This is referring to the actual encoding sizes rather than the native sizes.  These are mostly used for various fields in the AST structure, whereas constant values tend to be encoded as a vbr6, which is capable of larger sizes.  That said, it's not definitive, but I've also never had this fail in the half-dozen different bitcode files I've tested it on.  The comment is a little dated though because I'll be switching to `Word64#` representation in an upcoming PR.

Reply:
Ah, I see. Since this will change in an upcoming PR, I'll let this be.